### PR TITLE
[docs] best practices for upgrading zenml server

### PR DESCRIPTION
**Pull request by wjayesh**
## Describe changes
I added a new section: how-to/manage-zenml-server  with best practices around upgrading and using the zenml server.
- a new guide on how to use ZenML server in production with all the information that we have personally from deploying and managing ZenML Pro.
- a new guide on how to navigate upgrades of ZenML server with regards to user pipelines and artifacts
- general best practices for upgrading zenml
- removed the zenml deploy command and moved the migrations guide to this section
- subsumed the existing upgrade zenml pages into this new section

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+While upgrading ZenML is generally a smooth process, there are some best practices that you should follow to ensure a successful upgrade. Based on experiences shared by ZenML users, here are some key strategies and considerations.
+
+## Upgrading your server
+
+Follow the tips below while upgrading your server to mitigate data losses, downtime and other issues.
+
+### Data Backups
+
+- **Database Backup**: Before upgrading, create a backup of your MySQL database. This allows you to rollback if necessary.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:11]: maybe give a bash command for this?

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+### Upgrade Strategies
+
+- **Staged Upgrade**: For large organizations or critical systems, consider using two ZenML server instances (old and new) and migrating services one by one to the new version.
+    ![Server Migration Step 1](../../.gitbook/assets/server_migration_1.png)
+
+    ![Server Migration Step 2](../../.gitbook/assets/server_migration_2.png)
+
+- **Team Coordination**: If multiple teams share a ZenML server instance, coordinate the upgrade timing to minimize disruption.
+- **Separate Tenants**: Coordination between teams might be difficult if one team requires new features but the other can't upgrade yet. In such cases, it is recommended to use dedicated ZenML server instances per team or product to allow for more flexible upgrade schedules.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:24]: maybe a bit confusing. you use tenants vs 'server instances' somewhat interchangeably. you should be clear whether this section is for OSS deployments or ZenML-managed Pro tenants?

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+### Testing and Compatibility
+
+- **Local Testing**: It's a good idea to update your local server first and run some old pipelines to check for compatibility issues between the old and new versions.
+- **End-to-End Testing**: You can also develop simple end-to-end tests to ensure that the new version works with your pipeline code and your stack. ZenML already has an [extensive test suite](https://github.com/zenml-io/zenml/tree/main/tests) that we use for releases and you can use it as an example.
+- **Artifact Compatibility**: Be cautious with pickle-based [materializers](../../how-to/handle-data-artifacts/handle-custom-data-types.md), as they can be sensitive to changes in Python versions or libraries. Consider using version-agnostic materialization methods for critical artifacts.
+
+### Dependency Management
+
+- **Python Version**: Make sure that the Python version you are using is compatible with the ZenML version you are upgrading to.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:44]: Link to where they could check that

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+- **External Dependencies**: Be mindful of external dependencies (e.g. from integrations) that might be incompatible with the new version of ZenML. This could be the case when some older versions are no longer supported or maintained and the ZenML integration is updated to use a newer version.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:45]: link to release notes or tell people where they can see what the new dependency versions are for their integrations..

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+### Handling API Changes
+
+While ZenML strives for backward compatibility, be prepared for occasional breaking changes (e.g., the Pydantic 2 upgrade).
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:49]: link to the release notes / blog around the pydantic 2 upgrade

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+- **Changelog Review**: Always review the [changelog from new releases](https://github.com/zenml-io/zenml/releases) for new syntax, instructions, or breaking changes.
+- **Migration Scripts**: Use provided [migration scripts](../../how-to/manage-the-zenml-server/migration-guide/migration-guide.md) when available to handle database schema changes.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:52]: this is maybe a bit confusing. this only applies to really old versions?

File: docs/book/how-to/manage-the-zenml-server/upgrade-zenml-server.md [hunk 7]
```diff
@@ -4,24 +4,11 @@ description: Learn how to upgrade your server to a new version of ZenML for the
...
-* Run the deploy command again with this config file:
-
-  ```bash
-  zenml deploy --config=/PATH/TO/FILE
-  ```
-
-Any database schema updates are automatically handled by ZenML and unless mentioned otherwise, all of your data is
-migrated to the new version, intact.
-{% endtab %}
+In general, upgrade your ZenML server as soon as you can once a new version is released. New versions come with a lot of improvements and fixes that you can benefit from.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/upgrade-zenml-server.md:9]:
```suggestion
In general, upgrade your ZenML server as soon as you can once a new version is released. New versions come with a lot of improvements and fixes from which you can benefit.
```

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
+---
+description: >
+  Learn about best practices for using ZenML server in production environments.
+---
+
+# Using ZenML server in production
+
+Setting up a ZenML server is pretty straightforward. However, most people have to move ahead of day zero operations and in such cases, it helps to learn best practices around setting up your ZenML server in a production-ready way. This guide encapsulates all the tips and tricks we've learned ourselves and from working with people who use ZenML in production environments. Following are some of the best practices we recommend.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md:8]:
```suggestion
Setting up a ZenML server is pretty straightforward. However, most people have to move ahead of so-called 'day zero' operations and in such cases, it helps to learn best practices around setting up your ZenML server in a production-ready way. This guide encapsulates all the tips and tricks we've learned ourselves and from working with people who use ZenML in production environments. Following are some of the best practices we recommend.
```

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+{% hint style="info" %}
+If you are using ZenML Pro, you don't have to worry about any of these. We have got you covered!
+You can sign up for a free trial [here](https://cloud.zenml.io).
+{% endhint %}
+
+## Autoscaling replicas
+
+In production, you often have to run bigger and longer running pipelines that might strain your server's resources. It is a good idea to set up autoscaling for your ZenML server so that you don't have to worry about your pipeline runs getting interrupted or your Dashboard slowing down due to high traffic.
+
+How you do it depends greatly on the environment you have deployed your ZenML server in. Below are some commmon deployment options and how to set up autoscaling for them.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md:19]:
```suggestion
How you do it depends greatly on the environment in which you have deployed your ZenML server. Below are some common deployment options and how to set up autoscaling for them.
```

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+By default, it is set to 40. If you are using any other deployment option, you can set the `ZENML_SERVER_THREAD_POOL_SIZE` environment variable to the desired value.
+
+Once this is set, you should also modify the `zenml.database.poolSize` and `zenml.database.maxOverflow` values to ensure that the ZenML server workers do not block on database connections (i.e. the sum of the pool size and max overflow should be greater than or equal to the thread pool size). If you manage your own database, ensure these values are set appropriately.
+
+
+## Scaling the backing database
+
+An important component of the ZenML server deployment is the backing database. When you start scaling your ZenML server instances, you will also need to scale your database to avoid any bottlenecks.
+
+We would recommend starting out with a simple database instance and then monitoring it to decide if it needs scaling. Some common metrics to look out for:
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md:96]:
```suggestion
We would recommend starting out with a simple (single) database instance and then monitoring it to decide if it needs scaling. Some common metrics to look out for:
```

**Review by strickvl — CHANGES_REQUESTED** [review]

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md (hunk 1 above)
Comment (htahir1) [reply docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:11]: tbh as its so individual for every type of MySQL deployment, I would rather disagree and say don't give a bash command

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+- **Database Backup**: Before upgrading, create a backup of your MySQL database. This allows you to rollback if necessary.
+- **Automated Backups**: Consider setting up automatic daily backups of your database for added security. Most managed services like AWS RDS, Google Cloud SQL, and Azure Database for MySQL offer automated backup options.
+
+![Screenshot of backups in AWS RDS](../../.gitbook/assets/aws-rds-backups.png)
+
+### Upgrade Strategies
+
+- **Staged Upgrade**: For large organizations or critical systems, consider using two ZenML server instances (old and new) and migrating services one by one to the new version.
+    ![Server Migration Step 1](../../.gitbook/assets/server_migration_1.png)
```
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:19]: i had to fix this in your other PR. if you do a bullet point images format them with newlines and indent them in like this: 

```suggestion
- **Staged Upgrade**: For large organizations or critical systems, consider using two ZenML server instances (old and new) and migrating services one by one to the new version.

![Server Migration Step 1](../../.gitbook/assets/server_migration_1.png)

```

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md (hunk 2 above)
Comment (htahir1) [reply docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:24]: Yes, in this case id just say server always

Comment (htahir1) [reply docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:24]: Here you also have an opportunity to say that Zenml Pro makes this easy by having multi-tenant deployments of zenml

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+- **Avoid Mid-Pipeline Upgrades**: Be cautious of automated upgrades or redeployments that might interrupt long-running pipelines.
+
+## Upgrading your code
+
+Sometimes, you might have to upgrade your code to work with a new version of ZenML. This is true especially when you are moving from a really old version to a new major version. The following tips might help, in addition to everything you've learnt in this document so far.
+
+### Testing and Compatibility
+
+- **Local Testing**: It's a good idea to update your local server first and run some old pipelines to check for compatibility issues between the old and new versions.
```
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:38]: I wouldnt confuse it and say "local server". Just say , test it locally with the default stack, with `pip install zenml --upgrade` ?

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+- **End-to-End Testing**: You can also develop simple end-to-end tests to ensure that the new version works with your pipeline code and your stack. ZenML already has an [extensive test suite](https://github.com/zenml-io/zenml/tree/main/tests) that we use for releases and you can use it as an example.
+- **Artifact Compatibility**: Be cautious with pickle-based [materializers](../../how-to/handle-data-artifacts/handle-custom-data-types.md), as they can be sensitive to changes in Python versions or libraries. Consider using version-agnostic materialization methods for critical artifacts.
```
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:40]: Maybe ask people to try and load older artifacts after upgrading to ensure things go smoothly?

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md (hunk 6 above)
Comment (htahir1) [reply docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:52]: IMO he linked to the right thing and said if it applies. Makes sense to me

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md (hunk 8 above)
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md:8]: Dont say "its striaghtforward" and then have a huge guide :-) no need to take emotional opinions I think

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+{% endtab %}
+{% endtabs %}
+
+## Backups
+
+The data in your ZenML server is critical as it contains your pipeline runs, stack configurations, and other important information. It is, therefore, recommended to have a backup strategy in place to avoid losing any data.
+
+Some common strategies include:
+- Setting up automated backups with a good retention period (say 30 days).
```
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md:187]:
```suggestion
Some common strategies include:

- Setting up automated backups with a good retention period (say 30 days).
```

File: docs/book/toc.md [hunk 15]
```diff
@@ -12,8 +12,6 @@
   * [Deploy using HuggingFace Spaces](getting-started/deploying-zenml/deploy-using-huggingface-spaces.md)
   * [Deploy with custom images](getting-started/deploying-zenml/deploy-with-custom-image.md)
   * [Manage deployed services](getting-started/deploying-zenml/manage-the-deployed-services/README.md)
```
Comment (htahir1) [inline docs/book/toc.md:14]: What happens to this section now?

**Review by htahir1 — CHANGES_REQUESTED** [review]

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md (hunk 2 above)
Comment (wjayesh) [reply docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:24]: good points, i'll add it

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md (hunk 13 above)
Comment (wjayesh) [reply docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:42]: added that and a code block

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md (hunk 8 above)
Comment (wjayesh) [reply docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md]: ah, i agree. i wanted it to mean that just deploying vanilla zenml server is easy but then managing and maintaining it is hard. let me rephrase that

File: docs/book/toc.md (hunk 15 above)
Comment (wjayesh) [reply docs/book/toc.md:14]: i was thinking
- the troubleshoot stack components can go into the Manage stacks and componets section
- the secret store thing can go inside the secret management section.
wdyt?

File: .gitbook.yaml [hunk 16]
```diff
@@ -4,5 +4,12 @@ structure:
...
   summary: toc.md
 
-#redirects:
-#  help: ./support.md
\ No newline at end of file
+redirects:
+ reference/migration-guide/README.md: how-to/manage-the-zenml-server/migration-guide/migration-guide.md
+ reference/migration-guide/migration-zero-twenty.md: how-to/manage-the-zenml-server/migration-guide/migration-zero-twenty.md
+ reference/migration-guide/migration-zero-thirty.md: how-to/manage-the-zenml-server/migration-guide/migration-zero-thirty.md
+ reference/migration-guide/migration-zero-forty.md: how-to/manage-the-zenml-server/migration-guide/migration-zero-forty.md
```
Comment (htahir1) [inline .gitbook.yaml:11]: After the md needed on the left side or right side? Not sure.. can you check if  this works by syncing to a Gitbook space

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+![Screenshot of backups in AWS RDS](../../.gitbook/assets/aws-rds-backups.png)
+
+### Upgrade Strategies
+
+- **Staged Upgrade**: For large organizations or critical systems, consider using two ZenML server instances (old and new) and migrating services one by one to the new version.
+
+![Server Migration Step 1](../../.gitbook/assets/server_migration_1.png)
+
+
+    ![Server Migration Step 2](../../.gitbook/assets/server_migration_2.png)
```
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:28]: wrong again. try syncing to gitbook and check it please :-) 

```suggestion
![Server Migration Step 2](../../.gitbook/assets/server_migration_2.png)
```

File: docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+
+The data in your ZenML server is critical as it contains your pipeline runs, stack configurations, and other important information. It is, therefore, recommended to have a backup strategy in place to avoid losing any data.
+
+Some common strategies include:
+
+- Setting up automated backups with a good retention period (say 30 days).
+- Periodically exporting the data to an external storage (e.g. S3, GCS, etc.).
+- Manual backups before upgrading your server to avoid any problems.
+
+
```
Comment (htahir1) [inline docs/book/how-to/manage-the-zenml-server/using-zenml-server-in-prod.md:192]:
```suggestion
```

**Review by htahir1 — CHANGES_REQUESTED** [review]
Ok almost there! Can you sync to gitbook and give me a link please? id like to read the tabs and verify the images work!

**Comment by wjayesh** [comment]
@htahir1 this is the using-zenml-in-prod page deployed on my gitbook space https://zenml-io.gitbook.io/jayeshml/how-to/manage-your-zenml-server/using-zenml-server-in-prod

**Review by htahir1 — APPROVED** [review]
Approved minus the conflicts

File: docs/book/toc.md (hunk 15 above)
Comment (wjayesh) [reply docs/book/toc.md]: implemented this

File: docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md [hunk 19]
```diff
@@ -16,12 +21,18 @@ Follow the tips below while upgrading your server to mitigate data losses, downt
...
+
+
+![Server Migration Step 2](../../.gitbook/assets/server_migration_2.png)
 
 - **Team Coordination**: If multiple teams share a ZenML server instance, coordinate the upgrade timing to minimize disruption.
-- **Separate Tenants**: Coordination between teams might be difficult if one team requires new features but the other can't upgrade yet. In such cases, it is recommended to use dedicated ZenML server instances per team or product to allow for more flexible upgrade schedules.
+- **Separate ZenML Servers**: Coordination between teams might be difficult if one team requires new features but the other can't upgrade yet. In such cases, it is recommended to use dedicated ZenML server instances per team or product to allow for more flexible upgrade schedules.
+
+{% hint style="info" %}
+ZenML Pro comes with multi-tenacy which makes it easy for you to have multiple ZenML servers and switch between them. This is useful if you have multiple teams and you want to upgrade them at different times.
```
Comment (strickvl) [inline docs/book/how-to/manage-the-zenml-server/best-practices-upgrading-zenml.md:34]:
```suggestion
ZenML Pro comes with multi-tenancy which makes it easy for you to have multiple ZenML servers and switch between them. This is useful if you have multiple teams and you want to upgrade them at different times.
```

**Review by strickvl — CHANGES_REQUESTED** [review]

**Review by strickvl — APPROVED** [review]
